sssd_test_framework/roles/ad.py

```diff
@@ -1733,43 +1735,57 @@ def add(self) -> GPO:
 
     def link(
         self,
-        op: str | None = "New",
         target: str | None = None,
-        args: list[str] | str | None = None,
+        enforced: bool | None = None,
+        disabled: bool | None = False,
+        order: int | None = 0,
     ) -> GPO:
         """
-        Link the group policy to the a target object inside the directory, a site, domain or an ou.
-
-        ..Note::
-            The New and Set cmdlets are identical. To modify an an existing link,
-            change the $op parameter to "Set", i.e. to disable 'Enforced'
+        Link the group policy to the target object inside the directory, a site, domain or an ou.
 
-            ou_policy.link("Set", args=["-Enforced No"])
-
-        :param op: Cmdlet operation, defaults to "New"
-        :type op: str, optional
         :param target: Group policy target
         :type target: str, optional
-        :param args: Additional arguments
-        :type args: list[str] | None, optional
+        :param enforced: Enforced the policy
+        :type enforced: bool, optional
+        :param disabled: Disable the policy
+        :type disabled: bool, optional
+        :param order: Order number
+        :type order: int, optional
         :return: Group policy object
         :rtype: GPO
         """
-        if args is None:
-            args = []
+        args: CLIBuilderArgs = {}
 
-        if isinstance(args, list):
-            args = " ".join(args)
-        elif args is None:
-            args = ""
+        if enforced is True:
+            args |= {"Enforce": (self.cli.option.PLAIN, "Yes")}
```

Comment:
MS docs says "Enforced" (-d at the end)

You can simplify the code a lot by taking advantage of the cli builder (same on other methods and in Samba)

```diff
diff --git a/sssd_test_framework/roles/ad.py b/sssd_test_framework/roles/ad.py
index 714843a..63df3c4 100644
--- a/sssd_test_framework/roles/ad.py
+++ b/sssd_test_framework/roles/ad.py
@@ -1738,7 +1738,7 @@ class GPO(BaseObject[ADHost, AD]):
         target: str | None = None,
         enforced: bool | None = None,
         disabled: bool | None = False,
-        order: int | None = 0,
+        order: int | None = None,
     ) -> GPO:
         """
         Link the group policy to the target object inside the directory, a site, domain or an ou.
@@ -1749,41 +1749,34 @@ class GPO(BaseObject[ADHost, AD]):
         :type enforced: bool, optional
         :param disabled: Disable the policy
         :type disabled: bool, optional
-        :param order: Order number
-        :type order: int, optional
+        :param order: Order number, defaults to None
+        :type order: int | None, optional
         :return: Group policy object
         :rtype: GPO
         """
-        args: CLIBuilderArgs = {}
-
-        if enforced is True:
-            args |= {"Enforce": (self.cli.option.PLAIN, "Yes")}
-        if enforced is False:
-            args |= {"Enforce": (self.cli.option.PLAIN, "No")}
-
-        if disabled is True:
-            args |= {"LinkEnabled": (self.cli.option.PLAIN, "No")}
-        else:
-            args |= {"LinkEnabled": (self.cli.option.PLAIN, "Yes")}
-
-        if order != 0:
-            args |= {"Order": (self.cli.option.PLAIN, order)}
-
         if target is None and self.target is None:
             self.target = "Default-First-Site-Name"
 
         if target is not None and self.target is None:
             self.target = target
 
+        args: CLIBuilderArgs = {
+            "Guid": (self.cli.option.VALUE, self.cn),
+            "Target": (self.cli.option.VALUE, self.target),
+            "Enforced": (self.cli.option.VALUE, "Yes" if enforced else "No"),
+            "LinkEnabled": (self.cli.option.VALUE, "Yes" if not disabled else "No"),
+            "Order": (self.cli.option.VALUE, order),
+        }
+
         # The cmdlets take the same arguments, but one is for new links and the other is for existing links.
         # This is combined to simplify gpo management.
         new_link = self.role.host.conn.run(
-            f'New-GPLink -Guid "{self.cn}" -Target "{self.target}" {" ".join(self.cli.args(args))}',
+            self.cli.command("New-GPLink", args),
             raise_on_error=False,
         )
         if new_link.rc != 0:
             self.role.host.conn.run(
-                f'Set-GPLink -Guid "{self.cn}" -Target "{self.target}" {" ".join(self.cli.args(args))}',
+                self.cli.command("Set-GPLink", args),
                 raise_on_error=False,
             )
```

Result code:

```python
    def link(
        self,
        target: str | None = None,
        enforced: bool | None = None,
        disabled: bool | None = False,
        order: int | None = None,
    ) -> GPO:
        """
        Link the group policy to the target object inside the directory, a site, domain or an ou.

        :param target: Group policy target
        :type target: str, optional
        :param enforced: Enforced the policy
        :type enforced: bool, optional
        :param disabled: Disable the policy
        :type disabled: bool, optional
        :param order: Order number, defaults to None
        :type order: int | None, optional
        :return: Group policy object
        :rtype: GPO
        """
        if target is None and self.target is None:
            self.target = "Default-First-Site-Name"

        if target is not None and self.target is None:
            self.target = target

        args: CLIBuilderArgs = {
            "Guid": (self.cli.option.VALUE, self.cn),
            "Target": (self.cli.option.VALUE, self.target),
            "Enforced": (self.cli.option.VALUE, "Yes" if enforced else "No"),
            "LinkEnabled": (self.cli.option.VALUE, "Yes" if not disabled else "No"),
            "Order": (self.cli.option.VALUE, order),
        }

        # The cmdlets take the same arguments, but one is for new links and the other is for existing links.
        # This is combined to simplify gpo management.
        new_link = self.role.host.conn.run(
            self.cli.command("New-GPLink", args),
            raise_on_error=False,
        )
        if new_link.rc != 0:
            self.role.host.conn.run(
                self.cli.command("Set-GPLink", args),
                raise_on_error=False,
            )

        return self
```

Reply:
But especially the enforced attribute has to be fixed

Reply:
Fixed now for both Samba and AD.